File: src/bvh.ts
// bvh.ts
import { type Vec3 } from './math';
import { type IHittable, type AABB, PRIMITIVE_STRIDE } from './primitives';

export interface BVHNode {
  min: Vec3; max: Vec3;
  leftFirst: number; // 参照リストの開始インデックス
  triCount: number;  // 要素数
}

// --- ヘルパー関数: AABB操作 ---

// 空のAABBを作成 (無限小)
function emptyAABB(): AABB {
  return {
    min: { x: Infinity, y: Infinity, z: Infinity },
    max: { x: -Infinity, y: -Infinity, z: -Infinity },
    center: { x: 0, y: 0, z: 0 }
  };
}

// AABBを拡張して他のAABBを含める
function growAABB(target: AABB, add: AABB) {
  target.min.x = Math.min(target.min.x, add.min.x);
  target.min.y = Math.min(target.min.y, add.min.y);
  target.min.z = Math.min(target.min.z, add.min.z);
  target.max.x = Math.max(target.max.x, add.max.x);
  target.max.y = Math.max(target.max.y, add.max.y);
  target.max.z = Math.max(target.max.z, add.max.z);
}

// AABBの表面積を計算 (Cost計算用)
function aabbArea(aabb: AABB): number {
  const w = aabb.max.x - aabb.min.x;
  const h = aabb.max.y - aabb.min.y;
  const d = aabb.max.z - aabb.min.z;
  // 無効なボックス(初期値)の場合は0を返す
  if (w < 0 || h < 0 || d < 0) return 0;
  return 2 * (w * h + h * d + d * w);
}

// SAH用のビン情報
interface Bin {
  bounds: AABB;
  count: number;
}

export class BVHBuilder {
  nodes: BVHNode[] = [];
  sortedPrims: IHittable[] = [];

  // 入力: 球の配列と三角形の配列（生データ）を受け取る
  build(primitives: IHittable[]): { bvhNodes: Float32Array<ArrayBuffer>, unifiedPrimitives: Float32Array<ArrayBuffer> } {
    this.nodes = [];
    this.sortedPrims = [...primitives];


    // --- 2. BVH構築 ---
    const root = { min: { x: 0, y: 0, z: 0 }, max: { x: 0, y: 0, z: 0 }, leftFirst: 0, triCount: this.sortedPrims.length };
    this.nodes.push(root);

    // 全体のAABB
    this.updateNodeBounds(0);

    // 再帰的に分割開始
    // ※ console.time で計測するとSAHの効果（ビルド時間増 vs 描画負荷減）が見えて面白いです
    console.time("BVH Build");
    this.subdivide(0);
    console.timeEnd("BVH Build");

    // --- 3. データパック ---

    // ノードデータ (Float32)
    const bvhNodeData = this.packNodes();

    // プリミティブの一括パック
    const unifiedData = new Float32Array(this.sortedPrims.length * PRIMITIVE_STRIDE);
    for (let i = 0; i < this.sortedPrims.length; i++) {
      // 各オブジェクトの pack メソッドを呼ぶだけ！
      this.sortedPrims[i].pack(unifiedData, i * PRIMITIVE_STRIDE);
    }

    return { bvhNodes: bvhNodeData, unifiedPrimitives: unifiedData };
  }


  updateNodeBounds(nodeIdx: number) {
    const node = this.nodes[nodeIdx];
    let first = true;
    for (let i = 0; i < node.triCount; i++) {
      const aabb = this.sortedPrims[node.leftFirst + i].getAABB();
      if (first) {
        node.min = { ...aabb.min };
        node.max = { ...aabb.max };
        first = false;
      } else {
        node.min.x = Math.min(node.min.x, aabb.min.x);
        node.min.y = Math.min(node.min.y, aabb.min.y);
        node.min.z = Math.min(node.min.z, aabb.min.z);
        node.max.x = Math.max(node.max.x, aabb.max.x);
        node.max.y = Math.max(node.max.y, aabb.max.y);
        node.max.z = Math.max(node.max.z, aabb.max.z);
      }
    }
  }

  // ★改善ポイント: Binning SAH による分割
  subdivide(nodeIdx: number) {
    const node = this.nodes[nodeIdx];

    // 要素数が少なければ分割終了（葉ノード確定）
    if (node.triCount <= 4) return;

    // 現在のノードの大きさ
    const extent = {
      x: node.max.x - node.min.x,
      y: node.max.y - node.min.y,
      z: node.max.z - node.min.z,
    };

    // 最長軸を選択
    let axis = 0;
    if (extent.y > extent.x) axis = 1;
    if (extent.z > extent.x && extent.z > extent.y) axis = 2;

    const minVal = axis === 0 ? node.min.x : axis === 1 ? node.min.y : node.min.z;
    const length = axis === 0 ? extent.x : axis === 1 ? extent.y : extent.z;

    // 軸の長さがほぼゼロ（全ポリゴンが同じ位置など）なら分割不能なので終了
    if (length < 1e-6) return;

    // --- 1. Binning (ビニング) ---
    const BINS = 16; // ビン数 (増やしすぎるとビルドが遅くなる。16程度が丁度いい)
    const bins: Bin[] = [];
    for (let i = 0; i < BINS; i++) {
      bins.push({ bounds: emptyAABB(), count: 0 });
    }

    // 全プリミティブをビンに振り分け
    for (let i = 0; i < node.triCount; i++) {
      const prim = this.sortedPrims[node.leftFirst + i];
      const center = prim.getAABB().center;
      const pos = axis === 0 ? center.x : axis === 1 ? center.y : center.z;

      let binIdx = Math.floor(((pos - minVal) / length) * BINS);
      if (binIdx >= BINS) binIdx = BINS - 1; // 誤差対策
      if (binIdx < 0) binIdx = 0;

      bins[binIdx].count++;
      growAABB(bins[binIdx].bounds, prim.getAABB());
    }

    // --- 2. SAH コスト計算 ---
    // 左側と右側の累積データを計算
    const leftArea = new Float32Array(BINS);
    const leftCount = new Float32Array(BINS);
    const rightArea = new Float32Array(BINS);
    const rightCount = new Float32Array(BINS);

    // Left Sweep
    let currentBox = emptyAABB();
    let currentCount = 0;
    for (let i = 0; i < BINS; i++) {
      if (bins[i].count > 0) {
        growAABB(currentBox, bins[i].bounds);
        currentCount += bins[i].count;
      }
      leftArea[i] = aabbArea(currentBox);
      leftCount[i] = currentCount;
    }

    // Right Sweep
    currentBox = emptyAABB();
    currentCount = 0;
    for (let i = BINS - 1; i >= 0; i--) {
      if (bins[i].count > 0) {
        growAABB(currentBox, bins[i].bounds);
        currentCount += bins[i].count;
      }
      rightArea[i] = aabbArea(currentBox);
      rightCount[i] = currentCount;
    }

    // ベストな分割位置を探す
    let bestCost = Infinity;
    let bestSplit = -1; // ビンのインデックス (bestSplit番目のビンの右で切る)

    // ビンの境目 (0番目～BINS-2番目の後ろ) で試行
    for (let i = 0; i < BINS - 1; i++) {
      // 片方が空になる分割は意味がないのでスキップ
      if (leftCount[i] === 0 || rightCount[i + 1] === 0) continue;

      // SAH Cost = LeftArea * LeftCount + RightArea * RightCount
      const cost = leftArea[i] * leftCount[i] + rightArea[i + 1] * rightCount[i + 1];

      if (cost < bestCost) {
        bestCost = cost;
        bestSplit = i;
      }
    }

    // 良い分割が見つからなかった場合（コスト改善しない、または片寄る）は葉ノードにする
    if (bestSplit === -1) {
      return;
    }

    // --- 3. パーティショニング (並び替え) ---
    // クイックソートの Partition のような処理
    // bestSplit 以下のビンに入るものを左へ、それ以外を右へ集める
    let i = node.leftFirst;
    let j = i + node.triCount - 1;

    while (i <= j) {
      // 左側にあるべきものを探してスキップ
      const primI = this.sortedPrims[i];
      const centerI = primI.getAABB().center;
      const posI = axis === 0 ? centerI.x : axis === 1 ? centerI.y : centerI.z;
      let binIdxI = Math.floor(((posI - minVal) / length) * BINS);
      if (binIdxI >= BINS) binIdxI = BINS - 1;
      if (binIdxI < 0) binIdxI = 0;

      if (binIdxI <= bestSplit) {
        i++;
      } else {
        // 右側にあるべきものを見つけたので、右側から交換候補を探す
        const primJ = this.sortedPrims[j];
        const centerJ = primJ.getAABB().center;
        const posJ = axis === 0 ? centerJ.x : axis === 1 ? centerJ.y : centerJ.z;
        let binIdxJ = Math.floor(((posJ - minVal) / length) * BINS);
        if (binIdxJ >= BINS) binIdxJ = BINS - 1;
        if (binIdxJ < 0) binIdxJ = 0;

        if (binIdxJ > bestSplit) {
          j--;
        } else {
          // 交換 (Swap)
          const temp = this.sortedPrims[i];
          this.sortedPrims[i] = this.sortedPrims[j];
          this.sortedPrims[j] = temp;
          i++;
          j--;
        }
      }
    }

    // 左側の子の要素数
    const leftCountReal = i - node.leftFirst;

    // 安全策: 万が一どちらかが0個になってしまったら分割失敗として終了
    if (leftCountReal === 0 || leftCountReal === node.triCount) return;

    // 子ノードの作成
    const leftChildIdx = this.nodes.length;
    // 左の子
    this.nodes.push({
      min: { x: 0, y: 0, z: 0 }, max: { x: 0, y: 0, z: 0 },
      leftFirst: node.leftFirst,
      triCount: leftCountReal
    });
    // 右の子
    this.nodes.push({
      min: { x: 0, y: 0, z: 0 }, max: { x: 0, y: 0, z: 0 },
      leftFirst: i,
      triCount: node.triCount - leftCountReal
    });

    // 現在のノードを内部ノード化
    node.leftFirst = leftChildIdx;
    node.triCount = 0;

    // 再帰処理 (AABBを再計算してから潜る)
    this.updateNodeBounds(leftChildIdx);
    this.updateNodeBounds(leftChildIdx + 1);
    this.subdivide(leftChildIdx);
    this.subdivide(leftChildIdx + 1);
  }

  packNodes() {
    const data = new Float32Array(this.nodes.length * 8);
    for (let i = 0; i < this.nodes.length; i++) {
      const n = this.nodes[i];
      const off = i * 8;
      data[off] = n.min.x; data[off + 1] = n.min.y; data[off + 2] = n.min.z; data[off + 3] = n.leftFirst;
      data[off + 4] = n.max.x; data[off + 5] = n.max.y; data[off + 6] = n.max.z; data[off + 7] = n.triCount;
    }
    return data;
  }
}

File: rust-shader-tools/src/bvh.rs
// src/bvh.rs
use crate::primitives::AABB;
use glam::{Mat4, Vec3, vec3};

#[derive(Clone, Copy, Debug, Default)]
pub struct BVHNode {
    pub aabb: AABB,
    pub left_first: u32,
    pub tri_count: u32,
}

#[derive(Clone, Copy, Debug, Default)]
struct Bin {
    bounds: AABB,
    count: u32,
}

impl Bin {
    fn new() -> Self {
        Self {
            bounds: AABB::empty(),
            count: 0,
        }
    }
}

pub struct BVHBuilder<'a> {
    pub nodes: Vec<BVHNode>,
    vertices: &'a [f32],
    indices: &'a [u32],
    tri_indices: Vec<usize>,
    tri_aabbs: Vec<AABB>,
    tri_centers: Vec<Vec3>,
}

impl<'a> BVHBuilder<'a> {
    pub fn new(vertices: &'a [f32], indices: &'a [u32]) -> Self {
        let tri_count = indices.len() / 3;
        let mut tri_aabbs = Vec::with_capacity(tri_count);
        let mut tri_centers = Vec::with_capacity(tri_count);
        let mut tri_indices = Vec::with_capacity(tri_count);

        for i in 0..tri_count {
            tri_indices.push(i);
            let i0 = indices[i * 3] as usize;
            let i1 = indices[i * 3 + 1] as usize;
            let i2 = indices[i * 3 + 2] as usize;

            let get_v = |idx: usize| {
                let b = idx * 4;
                vec3(vertices[b], vertices[b + 1], vertices[b + 2])
            };
            let v0 = get_v(i0);
            let v1 = get_v(i1);
            let v2 = get_v(i2);

            let min = v0.min(v1).min(v2);
            let max = v0.max(v1).max(v2);
            let size = max - min;
            let eps = 1e-5;
            let pad = vec3(
                if size.x < eps { eps } else { 0. },
                if size.y < eps { eps } else { 0. },
                if size.z < eps { eps } else { 0. },
            );

            let aabb = AABB {
                min: min - pad * 0.5,
                max: max + pad * 0.5,
            };
            tri_aabbs.push(aabb);
            tri_centers.push(aabb.center());
        }

        Self {
            nodes: Vec::new(),
            vertices,
            indices,
            tri_indices,
            tri_aabbs,
            tri_centers,
        }
    }

    pub fn build(&mut self) -> (Vec<f32>, Vec<u32>) {
        self.nodes.clear();
        let root = BVHNode {
            left_first: 0,
            tri_count: self.tri_indices.len() as u32,
            ..Default::default()
        };
        self.nodes.push(root);
        self.update_node_bounds(0);
        self.subdivide(0);

        let packed_nodes = self.pack_nodes();
        let mut sorted_indices = Vec::with_capacity(self.indices.len());
        for &tri_idx in &self.tri_indices {
            let base = tri_idx * 3;
            sorted_indices.push(self.indices[base]);
            sorted_indices.push(self.indices[base + 1]);
            sorted_indices.push(self.indices[base + 2]);
        }
        (packed_nodes, sorted_indices)
    }
    pub fn build_with_ids(&mut self) -> (Vec<f32>, Vec<u32>, Vec<usize>) {
        self.nodes.clear();
        let root = BVHNode {
            left_first: 0,
            tri_count: self.tri_indices.len() as u32,
            ..Default::default()
        };
        self.nodes.push(root);
        self.update_node_bounds(0);
        self.subdivide(0);

        let packed_nodes = self.pack_nodes();
        let mut sorted_indices = Vec::with_capacity(self.indices.len());
        for &tri_idx in &self.tri_indices {
            let base = tri_idx * 3;
            sorted_indices.push(self.indices[base]);
            sorted_indices.push(self.indices[base + 1]);
            sorted_indices.push(self.indices[base + 2]);
        }

        // 最後にこれも返す
        (packed_nodes, sorted_indices, self.tri_indices.clone())
    }

    pub fn root_aabb(&self) -> AABB {
        if self.nodes.is_empty() {
            AABB::empty()
        } else {
            self.nodes[0].aabb
        }
    }

    fn update_node_bounds(&mut self, node_idx: usize) {
        let node = self.nodes[node_idx];
        let mut aabb = AABB::empty();
        for i in 0..node.tri_count {
            let tri_id = self.tri_indices[(node.left_first + i) as usize];
            aabb = aabb.union(&self.tri_aabbs[tri_id]);
        }
        self.nodes[node_idx].aabb = aabb;
    }

    fn subdivide(&mut self, node_idx: usize) {
        let node = self.nodes[node_idx];
        if node.tri_count <= 4 {
            return;
        }

        let extent = node.aabb.max - node.aabb.min;
        let axis = if extent.y > extent.x {
            1
        } else if extent.z > extent.x && extent.z > extent.y {
            2
        } else {
            0
        };
        let split_len = match axis {
            0 => extent.x,
            1 => extent.y,
            _ => extent.z,
        };
        let split_min = match axis {
            0 => node.aabb.min.x,
            1 => node.aabb.min.y,
            _ => node.aabb.min.z,
        };

        if split_len < 1e-6 {
            return;
        }

        const BINS: usize = 16;
        let mut bins = [Bin::new(); BINS];
        let first = node.left_first as usize;
        let count = node.tri_count as usize;

        for i in 0..count {
            let tri_id = self.tri_indices[first + i];
            let pos = match axis {
                0 => self.tri_centers[tri_id].x,
                1 => self.tri_centers[tri_id].y,
                _ => self.tri_centers[tri_id].z,
            };
            let mut bin_idx = (((pos - split_min) / split_len) * (BINS as f32)) as usize;
            if bin_idx >= BINS {
                bin_idx = BINS - 1;
            }
            bins[bin_idx].count += 1;
            bins[bin_idx].bounds = bins[bin_idx].bounds.union(&self.tri_aabbs[tri_id]);
        }

        let mut left_area = [0.; BINS];
        let mut left_count = [0; BINS];
        let mut right_area = [0.; BINS];
        let mut right_count = [0; BINS];
        let mut curr_box = AABB::empty();
        let mut curr_sum = 0;

        for i in 0..BINS {
            curr_sum += bins[i].count;
            curr_box = curr_box.union(&bins[i].bounds);
            left_area[i] = curr_box.area();
            left_count[i] = curr_sum;
        }
        curr_box = AABB::empty();
        curr_sum = 0;
        for i in (0..BINS).rev() {
            curr_sum += bins[i].count;
            curr_box = curr_box.union(&bins[i].bounds);
            right_area[i] = curr_box.area();
            right_count[i] = curr_sum;
        }

        let mut best_cost = f32::INFINITY;
        let mut best_split = usize::MAX;
        for i in 0..(BINS - 1) {
            if left_count[i] == 0 || right_count[i + 1] == 0 {
                continue;
            }
            let cost = left_area[i] * (left_count[i] as f32)
                + right_area[i + 1] * (right_count[i + 1] as f32);
            if cost < best_cost {
                best_cost = cost;
                best_split = i;
            }
        }

        if best_split == usize::MAX {
            return;
        }

        let mut i = first;
        let mut j = first + count - 1;
        while i <= j {
            let get_bin = |tri_id: usize| {
                let pos = match axis {
                    0 => self.tri_centers[tri_id].x,
                    1 => self.tri_centers[tri_id].y,
                    _ => self.tri_centers[tri_id].z,
                };
                let mut idx = (((pos - split_min) / split_len) * (BINS as f32)) as usize;
                if idx >= BINS {
                    idx = BINS - 1;
                }
                idx
            };

            let bin_i = get_bin(self.tri_indices[i]);

            if bin_i <= best_split {
                i += 1;
            } else if get_bin(self.tri_indices[j]) > best_split {
                if j == 0 {
                    break;
                }
                j -= 1;
            } else {
                self.tri_indices.swap(i, j);
                i += 1;
                if j == 0 {
                    break;
                }
                j -= 1;
            }
        }

        let left_count = i - first;
        if left_count == 0 || left_count == count {
            return;
        }

        let left_child_idx = self.nodes.len();
        self.nodes.push(BVHNode {
            left_first: first as u32,
            tri_count: left_count as u32,
            ..Default::default()
        });
        self.nodes.push(BVHNode {
            left_first: i as u32,
            tri_count: (count - left_count) as u32,
            ..Default::default()
        });

        self.nodes[node_idx].left_first = left_child_idx as u32;
        self.nodes[node_idx].tri_count = 0;

        self.update_node_bounds(left_child_idx);
        self.update_node_bounds(left_child_idx + 1);
        self.subdivide(left_child_idx);
        self.subdivide(left_child_idx + 1);
    }

    fn pack_nodes(&self) -> Vec<f32> {
        let mut data = vec![0.; self.nodes.len() * 8];
        for (i, node) in self.nodes.iter().enumerate() {
            let off = i * 8;
            data[off + 0] = node.aabb.min.x;
            data[off + 1] = node.aabb.min.y;
            data[off + 2] = node.aabb.min.z;
            data[off + 3] = node.left_first as f32;
            data[off + 4] = node.aabb.max.x;
            data[off + 5] = node.aabb.max.y;
            data[off + 6] = node.aabb.max.z;
            data[off + 7] = node.tri_count as f32;
        }
        data
    }
}

// --- TLAS / Instance ---

#[derive(Clone, Copy, Debug)]
#[repr(C)]
pub struct Instance {
    pub transform: Mat4,
    pub inverse_transform: Mat4,
    pub blas_node_offset: u32,
    pub attr_offset: u32,
    pub instance_id: u32,
    pub pad: u32,
}
impl Default for Instance {
    fn default() -> Self {
        Self {
            transform: Mat4::IDENTITY,
            inverse_transform: Mat4::IDENTITY,
            blas_node_offset: 0,
            attr_offset: 0,
            instance_id: 0,
            pad: 0,
        }
    }
}

// Reusing BVHNode structure for TLAS but different field meanings
// left_first -> left_right (child index or instance index)
// tri_count -> is_leaf (0 or 1)
pub type TLASNode = BVHNode;

pub struct TLASBuilder<'a> {
    pub nodes: Vec<TLASNode>,
    pub instances: &'a [Instance],
    pub instance_indices: Vec<usize>,
    pub instance_aabbs: Vec<AABB>,
    pub instance_centers: Vec<Vec3>,
}

impl<'a> TLASBuilder<'a> {
    pub fn new(instances: &'a [Instance], blas_aabbs: &[AABB]) -> Self {
        let count = instances.len();
        let mut instance_aabbs = Vec::with_capacity(count);
        let mut instance_centers = Vec::with_capacity(count);
        let instance_indices = (0..count).collect();

        for i in 0..count {
            // Note: In a real engine, we map instance.mesh_id -> blas_aabb.
            // Here assuming 1-to-1 or manually mapped slice passed in.
            let local_aabb = blas_aabbs[i];
            let world_aabb = local_aabb.transform(instances[i].transform);
            instance_aabbs.push(world_aabb);
            instance_centers.push(world_aabb.center());
        }

        Self {
            nodes: Vec::new(),
            instances,
            instance_indices,
            instance_aabbs,
            instance_centers,
        }
    }

    pub fn build(&mut self) -> (Vec<f32>, Vec<Instance>) {
        self.nodes.clear();
        if self.instances.is_empty() {
            return (vec![], vec![]);
        }

        let root = TLASNode {
            left_first: 0,
            tri_count: 1, // Start as leaf if count=1? or handle in subdivide
            ..Default::default()
        };
        self.nodes.push(root); // Dummy root

        // Setup Root Properties
        let mut root_aabb = AABB::empty();
        for aabb in &self.instance_aabbs {
            root_aabb = root_aabb.union(aabb);
        }
        self.nodes[0].aabb = root_aabb;
        self.nodes[0].tri_count = self.instances.len() as u32; // Temporary use of tri_count as "count"

        self.subdivide(0);

        let packed = self.pack_nodes();

        let mut sorted_instances = Vec::with_capacity(self.instances.len());
        for &idx in &self.instance_indices {
            sorted_instances.push(self.instances[idx]);
        }

        (packed, sorted_instances)
    }

    fn subdivide(&mut self, node_idx: usize) {
        let node = self.nodes[node_idx];
        let count = node.tri_count as usize; // Stored temporarily
        let first = node.left_first as usize; // Stored temporarily

        if count == 1 {
            // Leaf
            self.nodes[node_idx].tri_count = 1; // is_leaf = 1
            self.nodes[node_idx].left_first = first as u32; // Instance Index (in sorted list)
            return;
        }

        // Midpoint Split (Simple for TLAS)
        let extent = node.aabb.max - node.aabb.min;
        let axis = if extent.y > extent.x {
            1
        } else if extent.z > extent.x && extent.z > extent.y {
            2
        } else {
            0
        };

        // Sort indices based on center along axis
        let slice = &mut self.instance_indices[first..first + count];
        slice.sort_by(|&a, &b| {
            let ca = match axis {
                0 => self.instance_centers[a].x,
                1 => self.instance_centers[a].y,
                _ => self.instance_centers[a].z,
            };
            let cb = match axis {
                0 => self.instance_centers[b].x,
                1 => self.instance_centers[b].y,
                _ => self.instance_centers[b].z,
            };
            ca.partial_cmp(&cb).unwrap_or(std::cmp::Ordering::Equal)
        });

        let mid = count / 2;
        let left_count = mid;
        let right_count = count - mid;

        let left_child_idx = self.nodes.len();
        self.nodes.push(Default::default()); // Left
        self.nodes.push(Default::default()); // Right

        // Update Parent (Internal)
        self.nodes[node_idx].tri_count = 0; // is_leaf = 0
        self.nodes[node_idx].left_first = left_child_idx as u32; // Left Child Index

        // Setup Children
        // Left
        let mut left_aabb = AABB::empty();
        for i in 0..left_count {
            left_aabb = left_aabb.union(&self.instance_aabbs[self.instance_indices[first + i]]);
        }
        self.nodes[left_child_idx].aabb = left_aabb;
        self.nodes[left_child_idx].left_first = first as u32;
        self.nodes[left_child_idx].tri_count = left_count as u32;

        // Right
        let mut right_aabb = AABB::empty();
        for i in 0..right_count {
            right_aabb =
                right_aabb.union(&self.instance_aabbs[self.instance_indices[first + mid + i]]);
        }
        self.nodes[left_child_idx + 1].aabb = right_aabb;
        self.nodes[left_child_idx + 1].left_first = (first + mid) as u32;
        self.nodes[left_child_idx + 1].tri_count = right_count as u32;

        self.subdivide(left_child_idx);
        self.subdivide(left_child_idx + 1);
    }

    fn pack_nodes(&self) -> Vec<f32> {
        let mut data = vec![0.; self.nodes.len() * 8];
        for (i, node) in self.nodes.iter().enumerate() {
            let off = i * 8;
            data[off + 0] = node.aabb.min.x;
            data[off + 1] = node.aabb.min.y;
            data[off + 2] = node.aabb.min.z;
            data[off + 3] = node.left_first as f32; // Child Idx or Instance Idx
            data[off + 4] = node.aabb.max.x;
            data[off + 5] = node.aabb.max.y;
            data[off + 6] = node.aabb.max.z;
            data[off + 7] = node.tri_count as f32; // is_leaf (0 or 1)
        }
        data
    }
}
